vervet-underground/internal/storage/mem/mem.go

```diff
@@ -6,81 +6,252 @@ package mem
+
+	// map keys automatically make versions unique rather than iterate through array for each entry
+	// service, versionSlice
+	for _, versionSlice := range s.serviceVersions {
+		// index, version
+		for _, version := range versionSlice {
+			uniqueVersions[version.String()] = version
+		}
 	}
-	s.contents[k][digest] = contents
-	s.serviceVersions[k] = append(s.serviceVersions[k], contentRevision{
-		timestamp: scrapeTime,
-		digest:    digest,
-	})
+
+	for _, value := range uniqueVersions {
+		s.collatedVersions = append(s.collatedVersions, value)
+	}
+
+	// must sort at end
+	sort.Sort(s.collatedVersions)
+}
+
+func (s *Storage) collateVersion(version vervet.Version) error {
+	// number of services maximum needed
+	contentRevisionCollection := make([]ContentRevision, 0)
+
+	// preprocessing all relevant docs in byte format before combining
+	for service, versionSlice := range s.serviceVersions {
+		// If there is an exact match on versions 1-to-1
+		if revisions, ok := s.serviceVersionMappedRevisionSpecs[service][version.String()]; ok {
+			// TODO: takes first revision for scaffolding REVISIT!!!!
+			for _, contentRevision := range revisions {
+				contentRevisionCollection = append(contentRevisionCollection, contentRevision)
+				break
+			}
+		}
+
+		// If there is a fuzzy match on version supplied at collation for this service
+		// aka execute binarySearch on this service's available versions
+		resolvedVersion, err := versionSlice.Resolve(version)
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not resolve for service %s version %s", service, version.String())
+			continue
+		}
+
+		if revisions, ok := s.serviceVersionMappedRevisionSpecs[service][resolvedVersion.String()]; ok {
+			// TODO: takes first revision for scaffolding REVISIT!!!!
+			for _, contentRevision := range revisions {
+				contentRevisionCollection = append(contentRevisionCollection, contentRevision)
+				break
+			}
+		}
+	}
+
+	err := s.mergeContentRevisionCollection(version, contentRevisionCollection)
+	if err != nil {
+		log.Error().Err(err).Msg("Could not Merge specs for all services %s")
+		return err
+	}
+	return nil
+}
+
+func (s *Storage) mergeContentRevisionCollection(version vervet.Version, serviceRevisionCollection []ContentRevision) error {
+	loader := openapi3.NewLoader()
+	var dst *openapi3.T
+	for _, serviceRevision := range serviceRevisionCollection {
+		// JSON will deserialize here correctly
+		src, err := loader.LoadFromData(serviceRevision.blob)
+		if err != nil {
+			log.Error().
+				Err(err).
+				Msgf("Could not merge ServiceRevision %s:%s",
+					serviceRevision.serviceVersion,
+					serviceRevision.digest)
+			return err
+		}
+
+		if dst == nil {
+			dst = src
+		} else {
+			// TODO: evaluate whether to use replace bool or not during merging
+			vervet.Merge(dst, src, true)
+		}
+	}
+
+	// lock memory safely here
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.collatedVersionedSpecs[version] = *dst
```

Comment:
If you're doing something trivial with a mutex held, and there's no risk of early return or panic, you can:

```suggestion
	s.mu.Lock()
	s.collatedVersionedSpecs[version] = *dst
	s.mu.Unlock()
```